loguru_mypy/__init__.py

```diff
@@ -98,7 +101,7 @@ def _loguru_logger_call_handler(
     log_msg_expr = ctx.args[0][0]
     logger_opts = loggers.get(ctx.type) or DEFAULT_OPTS
 
-    if isinstance(log_msg_expr, StrExpr):
+    if isinstance(log_msg_expr, (StrExpr, NameExpr)):
```

Comment:
I actually been playing with this around #49 and I am pretty sure that `NameExpr` might resolve to anything not only `StrExp` which basically mean that if we encounter somethig non-trivial being:

- `NameExpr`
- `OpExpr` (i.e. `1+2`)

etc, we must resolve underlying type and maybe recall (as in recursively) `_loguru_logger_call_handler` function or postpone type checking (if that's a thing).

Reply:
You mean that in this situation

```python
from loguru import logger

foo = "bar"
logger.info(foo, 1)
```

We get a NameExpr but we would like to check it as if it was written like

```python
from loguru import logger

logger.info("bar", 1)
```

This was my initial reasoning as well. But the thing is NameExpr does not work like you would expect (at least not like I had expected). At first I tried to do something along the lines of:

```python
# <snip>
if isinstance(log_msg_expr, NameExpr):
  node = log_msg_expr.node
  # You would think that `node` is an Expression and you could cast it down to a StrExpr like
  if isinstance(node, StrExpr):
    _check_str_format_call(node, ctx)
  # But no, node is an Optional[SymbolNode]. So what you can do is
  if isinstance(node, Var):
    # ... Except a Var holds nothing that resembles an Expression.
    # Okay... Wait. I see something here.. What if I get the
    value = node.final_value
    # then build a
    str_expr = StrExpr(value=value)
    # and then... oh no. `node.final_value` is None.
```

In fact, the only time when `.final_value` is not None is when the variable is `Final`. (Which makes sense: we're talking about a *variable* after all.)  So after some trial and error I *did* end up with working code using this approach that would correctly handle something like this:

```python
from loguru import logger
from typing import Final

foo: Final = "bar"
logger.info(foo, 1)  # error
```

(Note that I even made this into a test to make sure the behaviour was the same.) But then I thought "that's pretty complicated and not very satisfying, does mypy even do that for `format`?"

```python
"bar".format(1)  # error
foo = "bar"
foo.format(1)  # ¯\_(ツ)_/¯

from typing import Final
foo: Final = "bar"
foo.format(1)  # error
```
"Welp, turns out it does... Wait... It *does*?" and that's when I realised I could just pass it all directly to mypy.

---

As for OpExpr, I have yet to dig into that but mypy doesn't know how to check

```python
("foo" + "bar").format(1)
```

so I'm pretty sure we'll be missing information just the same. (And I'm not sure it's worth the trouble as well, I mean who in their right mind writes `logger.debug("hello" + "{}", "world")`)

Reply:
That is fairly good point and I totally agree that there has to be a point where plugin says "I have no idea".
I mean some parts are obviously easy to interpret like for example `log.info(<StrExpr>)` whereas things like `OpExpr` or `NameExpr` are not trivial. 

For now, I would bet on ignoring `NameExpr` similarly to `IntExpr` and `FloatExpr`, so just add another thingy to the tuple and be done with it. As long as there is no a reliable way to find out what is there behind said expressions there's little that can be done.

So my bed is to have a mod here https://github.com/kornicameister/loguru-mypy/blob/master/loguru_mypy/__init__.py#L103 that does: `IntExpr`, `FloatExpr` and `NameExpr`.

If you agree, let's make this change and I will create separate issues to try and resolve a magic behind non trivial cases.
Thanks for that investigation, impressive job.

Reply:
I don't understand: mypy *does* support a NameExpr as the left-side of `.format` (as highlighted by the tests I added), it works.

Why would we ignore a case that *can* be easily validated?

If you're afraid of what happens if the NameExpr refers to a variable that is not a str, I've added a test to show that mypy catches it before it reaches us.

Reply:
Not really, I mean the case `variables__mypy_cant_tell` makes it hard to accept this PR without a fight ;). 
It is so obvious that string there is wrong. Don't get this wrong, I love this work of yours. It is so impressive :D. 
But one thing is that I wish to understand and another thing is to make it work. You have already done and provided a lot of useful knowledge but perhaps we are looking at this from wrong way. Since very beginning, this plugin aimed at capturing the exact call to `loguru.info` or `loguru.debug`. Maybe that is not correct in this very case. 

@sobolevn sorry for bothering you, but I believe you have worked much more with typing (at least in comparison to me). I wonder, if perhaps, you know if there is a way to resolve a type behind `NameExpr` so that we actually get another `Expr`. In case of this PR, that is going from `NameExpr` to `StrExpr`.

Anyway, apart from that, I will put a message to `Gitter` of `mypy`. Perhaps they will tell us something useful.

Reply:
Yes, you can always infer type of any variable: `ctx.api.expr_checker.accept` is the way to go. It accepts any `Expr` node and returns the type back.

Reply:
Example: https://github.com/dry-python/returns/blob/master/returns/contrib/mypy/_features/decorators.py#L29

Reply:
@sobolevn Thank you but I don't think `ExpressionChecker.accept` does what we want. We don't want a `Expression` ⟼ `Type`, we want a `NameExpr` ⟼ `Expression`. Again, for context we'd like to be able to lint something like this:

```python
foo = "foo {} baz"
foo.format()
# ^ This (`foo`) is a `NameExpr`
```

It's the *name* of a variable but we'd like to resolve that variable to the expression that defined it so that we can in turn analyse that expression to see if we can do some useful linting. In my example above, that would be returning `Cannot find replacement for positional format specifier 0` because the variable `foo` was in fact defined with a constant string expression and we know it's missing an argument. In other words, we'd like to turn the `NameExpr` «foo» into the `StrExpr` «"foo {} baz"» that was used to define the variable. As if we were rewriting the code into a canonical form.

``` python
"foo {} baz".format()
```

But looking at mypy's code, **I don't think that's possible**.

Note that I didn't choose my first example above at random, I used the builtin `.format()` method **which mypy already knows how to lint**. And if you try to lint my example **mypy won't give you any error**.

It doesn't feel right to me either, but from what I understand, there's nothing we can do about it, at least not with the way mypy currently works.

Reply:
> But looking at mypy's code, I don't think that's possible.

Yes, I agree with this conclusion.